### PR TITLE
Refactor shard modes logic

### PR DESCRIPTION
**Pull request by fyrchik**

### CODE REVIEW

File: cmd/neofs-cli/modules/control/shards_set_mode.go [hunk 1]
```diff
@@ -17,9 +17,10 @@ const (
 	shardIDFlag          = "id"
 	shardClearErrorsFlag = "clear-errors"
 
-	shardModeReadOnly  = "read-only"
-	shardModeReadWrite = "read-write"
-	shardModeDegraded  = "degraded"
+	shardModeReadOnly         = "read-only"
+	shardModeReadWrite        = "read-write"
+	shardModeDegraded         = "degraded"
```
Comment (realloc) [inline cmd/neofs-cli/modules/control/shards_set_mode.go:22]: Let's rename this state to `degraded-read-write` for the sake of consistency.

File: pkg/local_object_storage/writecache/mode.go [hunk 2]
```diff
@@ -13,22 +13,28 @@ var ErrReadOnly = errors.New("write-cache is in read-only mode")
...
+		c.persistMemoryCache()
 	}
 
-	// Because modeMtx is taken no new objects will arrive an all other modifying
-	// operations are completed.
-	// 1. Persist objects already in memory on disk.
-	c.persistMemoryCache()
+	if c.db != nil {
+		if err := c.db.Close(); err != nil {
+			return err
```
Comment (carpawell) [inline pkg/local_object_storage/writecache/mode.go:34]: add some context to the err?

File: pkg/local_object_storage/writecache/mode.go [hunk 3]
```diff
@@ -13,22 +13,28 @@ var ErrReadOnly = errors.New("write-cache is in read-only mode")
...
 // and all background jobs are suspended.
-func (c *cache) SetMode(m mode.Mode) {
+func (c *cache) SetMode(m mode.Mode) error {
 	c.modeMtx.Lock()
 	defer c.modeMtx.Unlock()
-	if c.mode == m {
-		return
+
+	if m.ReadOnly() == c.readOnly() {
+		c.mode = m
```
Comment (carpawell) [inline pkg/local_object_storage/writecache/mode.go:21]: do we need that line?

File: cmd/neofs-cli/modules/control/shards_set_mode.go (hunk 1 above)
Comment (carpawell) [reply cmd/neofs-cli/modules/control/shards_set_mode.go:22]: @fyrchik

**Review by carpawell — COMMENTED** [review]
Did you check all the `CreateIfNotExists` in the metabase's "read" operation?

File: pkg/local_object_storage/writecache/mode.go [hunk 4]
```diff
@@ -3,20 +3,8 @@ package writecache
...
-const (
-	// ModeReadWrite is a default mode allowing objects to be flushed.
-	ModeReadWrite Mode = iota
-
-	// ModeReadOnly is a mode in which write-cache doesn't flush anything to a metabase.
-	ModeReadOnly
 
-	// ModeDegraded is similar to a shard's degraded mode.
-	ModeDegraded
+	"github.com/nspcc-dev/neofs-node/pkg/local_object_storage/shard/mode"
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/writecache/mode.go:7]: It depends on a subpackage of shard, in essense the mode is defined for shard, writecache inherits it.

File: pkg/local_object_storage/engine/delete.go [hunk 5]
```diff
@@ -72,6 +73,10 @@ func (e *StorageEngine) delete(prm DeletePrm) (DeleteRes, error) {
 
 			resExists, err := sh.Exists(existsPrm)
 			if err != nil {
+				_, ok := err.(*objectSDK.SplitInfoError)
+				if ok && shard.IsErrNotFound(err) && shard.IsErrRemoved(err) {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/engine/delete.go:77]: Fixed

File: pkg/local_object_storage/shard/exists.go [hunk 6]
```diff
@@ -35,27 +33,23 @@ func (p ExistsRes) Exists() bool {
...
+
+		var res blobstor.ExistsRes
+		res, err = s.blobStor.Exists(p)
+		exists = res.Exists()
+	} else {
+		var existsPrm meta.ExistsPrm
+		existsPrm.SetAddress(prm.addr)
+
+		var res meta.ExistsRes
+		res, err = s.metaBase.Exists(existsPrm)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/shard/exists.go:51]: The error is returned to the caller.

File: cmd/neofs-cli/modules/control/shards_set_mode.go (hunk 1 above)
Comment (fyrchik) [reply cmd/neofs-cli/modules/control/shards_set_mode.go:22]: Fixed

File: pkg/local_object_storage/writecache/mode.go (hunk 3 above)
Comment (fyrchik) [reply pkg/local_object_storage/writecache/mode.go:21]: Yes, we can change `Degraded` flag.

File: pkg/local_object_storage/writecache/mode.go (hunk 2 above)
Comment (fyrchik) [reply pkg/local_object_storage/writecache/mode.go]: Fixed

File: pkg/local_object_storage/engine/delete.go [hunk 7]
```diff
@@ -72,6 +73,10 @@ func (e *StorageEngine) delete(prm DeletePrm) (DeleteRes, error) {
 
 			resExists, err := sh.Exists(existsPrm)
 			if err != nil {
+				_, ok := err.(*objectSDK.SplitInfoError)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/engine/delete.go:76]: We can do it in a separate task. I think `engine.Delete` can accept a single value, it is just a `for` loop. Also makes metrics more meaningful.

File: pkg/local_object_storage/writecache/mode.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/writecache/mode.go:8]: ok, what do you think about the nested structure for the nested structs then: `engine/shard/writecache`, etc? not for that PR

File: pkg/local_object_storage/shard/exists.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/shard/exists.go:51]: just wanted to clarify if we change behavior consciously: we checked blobstore directly if the meta returned an err (and shard was in the `Degraded` mode) previously. now we check meta and do nothing in the error case

[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/shard/exists.go:51]: Yes, previously `Degraded` was for "fallback to blobstor", now it is "without the metabase"

File: pkg/local_object_storage/writecache/mode.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/writecache/mode.go:8]: I think both options (current and `engine/shard/writecache`) are good.

File: pkg/local_object_storage/shard/mode/mode.go [hunk 8]
```diff
@@ -31,3 +31,13 @@ func (m Mode) String() string {
 		return "DEGRADED"
 	}
 }
+
+// NoMetabase returns true iff m is operating without the metabase.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/shard/mode/mode.go:42]: What do you mean? Probably, fixed.

File: pkg/local_object_storage/engine/delete.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/engine/delete.go:76]: > what about that func but for split info error?
Let's do it later. We won't need it after we introduce statuses.

[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/local_object_storage/engine/delete.go:76]: https://github.com/nspcc-dev/neofs-node/issues/1616

**Comment by fyrchik** [comment]
> Did you check all the CreateIfNotExists in the metabase's "read" operation?

Could you elaborate a bit?

File: pkg/local_object_storage/shard/mode/mode.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/local_object_storage/shard/mode/mode.go:42]: yes, has been fixed

**Review by carpawell — APPROVED** [review]

**Comment by carpawell** [comment]
> Could you elaborate a bit?

@fyrchik, I remember that we could create some buckets even in "read" operations via `CreateBucketIfNotExists`. Guess it is not smth that we wanna do if we try to save the disk and do not write any data to it. But Im not sure, did not check.

**Comment by fyrchik** [comment]
@carpawell  bbolt returns an error if `CreateBucket` is used in a transaction that is not writeable.

**Review by carpawell — APPROVED** [review]

**Pull request merged**